shared_storage/documentscontract.cpp

```diff
@@ -0,0 +1,196 @@
+/*
+ * SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+ * Author: BogDan Vatra <bogdan.vatra@kdab.com>
+ * SPDX-License-Identifier: MIT
+*/
+#include "contentresolver.h"
+#include "documentscontract.h"
+
+namespace android {
+
+using namespace content;
+
+namespace provider {
+namespace DocumentsContract {
+
+QString documentId(const net::Uri &uri)
+{
+    return QJniObject::callStaticObjectMethod("android/provider/DocumentsContract",
+                                              "getDocumentId",
+                                              "(Landroid/net/Uri;)Ljava/lang/String;",
+                                              uri.object()).toString();
+}
+
+QString treeDocumentId(const net::Uri &uri)
+{
+    return QJniObject::callStaticObjectMethod("android/provider/DocumentsContract",
+                                              "getTreeDocumentId",
+                                              "(Landroid/net/Uri;)Ljava/lang/String;",
+                                              uri.object()).toString();
+}
+
+QString rootId(const net::Uri &uri)
+{
+    return QJniObject::callStaticObjectMethod("android/provider/DocumentsContract",
+                                              "getRootId",
+                                              "(Landroid/net/Uri;)Ljava/lang/String;",
+                                              uri.object()).toString();
+}
+
+net::Uri buildChildDocumentsUriUsingTree(const net::Uri &uri)
+{
+    return buildChildDocumentsUriUsingTree(uri, treeDocumentId(uri));
+}
+
+net::Uri buildChildDocumentsUriUsingTree(const net::Uri &uri, const QString &parentDocumentId)
+{
+    return QJniObject::callStaticObjectMethod("android/provider/DocumentsContract",
+                                              "buildChildDocumentsUriUsingTree",
+                                              "(Landroid/net/Uri;Ljava/lang/String;)Landroid/net/Uri;",
+                                              uri.object(),
+                                              QJniObject::fromString(parentDocumentId).object());
+
+}
+
+
+net::Uri buildDocumentUriUsingTree(const net::Uri &treeUri, const QString &documentId)
+{
+    return QJniObject::callStaticObjectMethod("android/provider/DocumentsContract",
+                                              "buildDocumentUriUsingTree",
+                                              "(Landroid/net/Uri;Ljava/lang/String;)Landroid/net/Uri;",
+                                              treeUri.object(),
+                                              QJniObject::fromString(documentId).object());
+}
+
+net::Uri copyDocument(const net::Uri &sourceDocumentUri, const net::Uri &targetParentDocumentUri)
+{
+    return QJniObject::callStaticObjectMethod("android/provider/DocumentsContract",
```

Comment:
What happens when these methods fail? Is the return value empty then, or something? In java, probably an exception is thrown, is there a nice way to forward these to the user?

Reply:
The value is empty and callStaticObjectMethod will clean the exception automagically.
I wanted that all QJniObject::xxxx to throw C++ exceptions but Qt folks didn't like it...